packages/mcfunction/src/parser/macro.ts

```diff
@@ -0,0 +1,153 @@
+import * as core from '@spyglassmc/core'
+import { localeQuote, localize } from '@spyglassmc/locales'
+import type { MacroChildNode, MacroKeyNode, MacroNode } from '../node/index.js'
+
+/**
+ * Parse a macro.
+ */
+export function macro(): core.Parser<MacroNode> {
+	return (
+		src: core.Source,
+		ctx: core.ParserContext,
+	): core.Result<MacroNode> => {
+		const start = src.cursor
+		const children: MacroChildNode[] = []
+
+		// Handle the starting '$'
+		if (src.peek() == '$' && src.peek(2) !== '$(') {
+			const ans: MacroChildNode = {
+				type: 'mcfunction:macro_child',
+				range: core.Range.create(start, start + 1),
+				options: { type: 'sign' },
+				value: '$',
+			}
+			src.skip()
+			children.push(ans)
+		}
+
+		// Handle the rest of the line
+
+		let beginning = src.cursor
+		let txt = src.readUntil('$', '\r', '\n')
+		do { // Do-while because we always want it to work the first time
+			let wasMacro = false
+			const child: MacroChildNode = parse(src, ctx, txt, beginning, wasMacro)
+			if (child.options.type === 'macro') {
+				wasMacro = true
+			}
+			// Handle a non-macro $ in the middle of the line
+			let subMacro: MacroChildNode | undefined
+			if (
+				child.options.type === 'other' && child.value?.indexOf('$') !== -1
+			) {
+				const parts: String[] = child.value?.split('$') ?? []
+				let other = ''
+				for (const part of parts) {
+					if (part.charAt(0) == '(') { // We've found our macro key!
+						subMacro = parse(
+							src,
+							ctx,
+							'$' + part,
+							beginning + other.length - 1,
+							wasMacro,
+						)
+						child.range.end -= part.length + 1
+					} else { // Keep it in the normal string
+						other += part + '$'
+					}
+				}
+				wasMacro = true
+				child.value = other.substring(0, other.length - 1) // Remove the last '$'
+			}
+			children.push(child)
+			if (subMacro) { // Sub-macro added here because colorizer
+				children.push(subMacro)
+			}
+			// Prepare for the next block
+			beginning = src.cursor
+			if (wasMacro) {
+				txt = src.readUntil('\r', '\n', '$')
+			} else {
+				txt = src.readUntil('\r', '\n', ')')
+			}
+		} while (txt.length > 0)
+
+		// No actual macro
+		if (children.length < 3) {
```

Comment:
This is a very hacky way to check whether there were macro keys. Consider filtering the children list and checking their type. This would presumably also fail for the line `$$(cmd)`, which is valid.